### PR TITLE
[feature]: Implemented Struct instead of passing in variable in CaptureTraffic and GetRecordConfig

### PR DESCRIPTION
**Pull request by Akash-Singh04**
## Related Issue
  - Info about Issue or bug

Closes: #1482 

#### Describe the changes you've made
Implemented Struct ```Record Options``` to pass params into ```Capture Traffic``` and ```GetRecordConfig``` instead of passing in variables, inorder to have better coding practice and less merge conflicts.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added
NIL
#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
NA
## Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

image

### CODE REVIEW

**Comment by Akash-Singh04** [comment]
@shivamsouravjha Kindly Review this PR

File: cmd/record.go [hunk 1]
```diff
@@ -293,9 +293,21 @@ func (r *Record) GetCmd() *cobra.Command {
...
+				BuildDelay:       buildDelay,
+				Ports:            ports,
+				Filters:          &filters,
+				EnableTele:       enableTele,
+				PassThroughHosts: passThrough,
+			}
 			r.logger.Debug("the ports are", zap.Any("ports", ports))
-			r.recorder.CaptureTraffic(path, proxyPort, appCmd, appContainer, networkName, delay, buildDelay, ports, &filters, enableTele, passThrough)
+			r.recorder.CaptureTraffic(params)
 			return nil
```
Comment (shivamsouravjha) [inline cmd/record.go:311]: instead of params some better name can be selected

File: pkg/service/record/record.go [hunk 2]
```diff
@@ -29,28 +29,30 @@ func NewRecorder(logger *zap.Logger) Recorder {
 	}
 }
 
-func (r *recorder) CaptureTraffic(path string, proxyPort uint32, appCmd, appContainer, appNetwork string, Delay uint64, buildDelay time.Duration, ports []uint, filters *models.TestFilter, enableTele bool, passThroughHosts []models.Filters) {
-
+func (r *recorder) CaptureTraffic(params TrafficCaptureParams) {
 	var ps *proxy.ProxySet
+
 	stopper := make(chan os.Signal, 1)
```
Comment (shivamsouravjha) [inline pkg/service/record/record.go:35]: params isn't a suitable name, any better name can be used

File: pkg/service/record/service.go [hunk 3]
```diff
@@ -6,6 +6,19 @@ import (
...
+	Delay            uint64
+	BuildDelay       time.Duration
+	Ports            []uint
+	Filters          *models.TestFilter
+	EnableTele       bool
+	PassThroughHosts []models.Filters
+}
 type Recorder interface {
-	CaptureTraffic(path string, proxyPort uint32, appCmd, appContainer, networkName string, Delay uint64, buildDelay time.Duration, ports []uint, filters *models.TestFilter, enableTele bool, passThroughHosts []models.Filters)
+	CaptureTraffic(params TrafficCaptureParams)
```
Comment (shivamsouravjha) [inline pkg/service/record/service.go:23]: instead of putting in service rather in models folder something can be used

File: pkg/service/record/record.go [hunk 4]
```diff
@@ -79,49 +81,49 @@ func (r *recorder) CaptureTraffic(path string, proxyPort uint32, appCmd, appCont
...
 				switch err {
 				case hooks.ErrInterrupted:
 					r.Logger.Info("keploy terminated user application")
 					return
 				case hooks.ErrCommandError:
 				case hooks.ErrUnExpected:
-					r.Logger.Warn("user application terminated unexpectedly hence stopping keploy, please check application logs if this behaviour is not expected")
+					r.Logger.Warn("user application terminated unexpectedly hence stopping keploy, please check application logs if this behavior is not expected")
 				case hooks.ErrDockerError:
 					stopApplication = true
```
Comment (shivamsouravjha) [inline pkg/service/record/record.go:117]: unnecessary change

File: pkg/service/record/record.go [hunk 5]
```diff
@@ -68,8 +70,8 @@ func (r *recorder) CaptureTraffic(path string, proxyPort uint32, appCmd, appCont
 	case <-stopper:
 		return
 	default:
-		// load the ebpf hooks into the kernel
-		if err := loadedHooks.LoadHooks(appCmd, appContainer, 0, ctx, filters); err != nil {
+		// Load the ebpf hooks into the kernel
+		if err := loadedHooks.LoadHooks(params.AppCmd, params.AppContainer, 0, ctx, params.Filters); err != nil {
```
Comment (shivamsouravjha) [inline pkg/service/record/record.go:74]: updating the variable names seems unnecessary as no linter or anything is affected

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

File: pkg/service/record/record.go (hunk 4 above)
Comment (Akash-Singh04) [reply pkg/service/record/record.go:117]: behaviour spelling is fixed in the above change

File: pkg/service/record/service.go (hunk 3 above)
Comment (Akash-Singh04) [reply pkg/service/record/service.go:23]: Refactored to capturetraffic.go in models

File: pkg/service/record/record.go (hunk 4 above)
Comment (shivamsouravjha) [reply pkg/service/record/record.go:117]: Both are acceptable spellings.

**Comment by shivamsouravjha** [comment]
it's a better practice to let the reviewer resolve the conversation after they're satisfied with the changes.

File: pkg/service/record/service.go (hunk 3 above)
Comment (shivamsouravjha) [reply pkg/service/record/service.go:23]: seems to be unrelated to comment as
1. Params variable isn't changed
2. Struct still present in service file

File: pkg/service/record/record.go (hunk 5 above)
Comment (Akash-Singh04) [reply pkg/service/record/record.go:74]: Only the cases of the variables have been changed to ensure consistency with the struct. Do you want them camelCased?

File: pkg/service/record/service.go (hunk 3 above)
Comment (Akash-Singh04) [reply pkg/service/record/service.go]: Kindly check now

File: pkg/service/record/record.go (hunk 4 above)
Comment (Akash-Singh04) [reply pkg/service/record/record.go:117]: Do you want the change reverted?

File: cmd/record.go (hunk 1 above)
Comment (Akash-Singh04) [reply cmd/record.go]: Refactored

File: pkg/service/record/record.go (hunk 2 above)
Comment (Akash-Singh04) [reply pkg/service/record/record.go:35]: Refactored to CaptureTrafficConfig

File: pkg/models/capturetraffic.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+	ProxyPort        uint32
+	AppCmd           string
+	AppContainer     string
+	AppNetwork       string
+	Delay            uint64
+	BuildDelay       time.Duration
+	Ports            []uint
+	Filters          *TestFilter
+	EnableTele       bool
+	PassThroughHosts []Filters
```
Comment (shivamsouravjha) [inline pkg/models/capturetraffic.go:16]: any reason to make this all global?

File: pkg/service/record/record.go (hunk 4 above)
Comment (shivamsouravjha) [reply pkg/service/record/record.go:117]: yes can be reverted as someone might update to the same in near future.

File: pkg/service/record/service.go [hunk 7]
```diff
@@ -1,11 +1,9 @@
...
 
 import (
-	"time"
-
 	"go.keploy.io/server/pkg/models"
 )
 
 type Recorder interface {
-	CaptureTraffic(path string, proxyPort uint32, appCmd, appContainer, networkName string, Delay uint64, buildDelay time.Duration, ports []uint, filters *models.TestFilter, enableTele bool, passThroughHosts []models.Filters)
+	CaptureTraffic(params models.TrafficCaptureParams)
```
Comment (shivamsouravjha) [inline pkg/service/record/service.go:8]: variable name can be updated here

File: pkg/models/capturetraffic.go (hunk 6 above)
Comment (Akash-Singh04) [reply pkg/models/capturetraffic.go:16]: In the future, if we would need to use the CaptureTraffic, it would be easier to access the global struct to easily pass the vars from command to service layer

**Comment by Akash-Singh04** [comment]
@shivamsouravjha Kindly check the review comments and let me know if any more changes are required

File: pkg/service/record/service.go (hunk 7 above)
Comment (Akash-Singh04) [reply pkg/service/record/service.go]: Updated

File: pkg/service/record/record.go (hunk 4 above)
Comment (Akash-Singh04) [reply pkg/service/record/record.go]: Reverted

File: pkg/models/capturetraffic.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package models
+
+import "time"
+
+type TrafficCaptureParams struct {
+	Path             string
+	ProxyPort        uint32
```
Comment (shivamsouravjha) [inline pkg/models/capturetraffic.go:7]: TrafficCaptureConfig would be better name

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

File: pkg/service/record/record.go [hunk 9]
```diff
@@ -29,28 +28,28 @@ func NewRecorder(logger *zap.Logger) Recorder {
 	}
 }
 
-func (r *recorder) CaptureTraffic(path string, proxyPort uint32, appCmd, appContainer, appNetwork string, Delay uint64, buildDelay time.Duration, ports []uint, filters *models.TestFilter, enableTele bool, passThroughHosts []models.Filters) {
+func (r *recorder) CaptureTraffic(captureTrafficConfig models.TrafficCaptureParams) {
```
Comment (shivamsouravjha) [inline pkg/service/record/record.go:31]: if recordOption format is followed this would be just options(captureTrafficConfig)

Comment (Akash-Singh04) [reply pkg/service/record/record.go]: Refactored

File: pkg/models/capturetraffic.go (hunk 8 above)
Comment (Akash-Singh04) [reply pkg/models/capturetraffic.go]: Refactored

**Comment by Akash-Singh04** [comment]
@shivamsouravjha Kindly check the review comments and let me know if any more changes are required

**Comment by shivamsouravjha** [comment]
all the changes LGTM. 

 Although would like to discuss if we should try to use this structure in`GetRecordConfig` as well?
 
 We've a structure that stores all the value and we pass it to capture traffic. Is it possible to refactor `GetRecordConfig` as well?
 
 
 
Please explore the possibilities of it in this PR as well. Adding this in the issue description.

File: cmd/record.go [hunk 10]
```diff
@@ -167,7 +167,19 @@ func (r *Record) GetCmd() *cobra.Command {
...
+				AppNetwork:       networkName,
+				Delay:            delay,
+				BuildDelay:       buildDelay,
+				Ports:            ports,
+				Filters:          &filters,
+				EnableTele:       enableTele,
+				PassThroughHosts: passThrough,
+			}, configPath)
 			if err != nil {
 				if err == errFileNotFound {
```
Comment (shivamsouravjha) [inline cmd/record.go:184]: instead of declaring it twice better to assign to a variable and use it in both options.

File: pkg/service/record/service.go [hunk 11]
```diff
@@ -1,14 +1,13 @@
...
 	"go.keploy.io/server/pkg/platform/telemetry"
 )
 
 type Recorder interface {
-	CaptureTraffic(path string, proxyPort uint32, appCmd, appContainer, networkName string, dirName string, Delay uint64, buildDelay time.Duration, ports []uint, filters *models.TestFilter, tcDB platform.TestCaseDB, tele *telemetry.Telemetry, passThroughHosts []models.Filters)
-	StartCaptureTraffic(path string, proxyPort uint32, appCmd, appContainer, networkName string, Delay uint64, buildDelay time.Duration, ports []uint, filters *models.TestFilter, enableTele bool, passThroughHosts []models.Filters)
+	StartCaptureTraffic(options models.RecordOptions)
+  	CaptureTraffic(options models.RecordOptions,  dirName string, tcDB platform.TestCaseDB, tele *telemetry.Telemetry)
+	
 }
```
Comment (shivamsouravjha) [inline pkg/service/record/service.go:13]: instead of using dirName and variables after that. Better to make another struct `RecordingEnvironment` as the reason for making struct was to reduce number of variables passed.

File: pkg/models/recordoptions.go
Comment (shivamsouravjha) [inline pkg/models/recordoptions.go:1]: file name should me made in camel cases.

File: cmd/test.go [hunk 12]
```diff
@@ -250,7 +249,27 @@ func (t *Test) GetCmd() *cobra.Command {
...
 			passThroughHosts := []models.Filters{}
 
-			err = t.getTestConfig(&path, &proxyPort, &appCmd, &tests, &appContainer, &networkName, &delay, &buildDelay, &ports, &apiTimeout, &globalNoise, &testsetNoise, &coverageReportPath, &withCoverage, configPath, &ignoreOrdering, &passThroughHosts)
+			err = t.getTestConfig(
+				models.TestConfigOptions{
+					Path:               &path,
+					ProxyPort:          &proxyPort,
+					AppCmd:             &appCmd,
+					Tests:              &tests,
+					AppContainer:       &appContainer,
```
Comment (shivamsouravjha) [inline cmd/test.go:258]: again instead of defining twice better to assign to a variable and pass it everywhere.

File: pkg/models/testconfigoptions.go [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+	BuildDelay         *time.Duration
+	PassThroughPorts   *[]uint
+	ApiTimeout         *uint64
+	GlobalNoise        *GlobalNoise
+	TestSetNoise       *TestsetNoise
+	CoverageReportPath *string
+	WithCoverage       *bool
+	ConfigPath         *string
+	IgnoreOrdering     *bool
+	PassThroughHosts   *[]Filters
```
Comment (shivamsouravjha) [inline pkg/models/testconfigoptions.go:22]: instead of redefining it again better to make a struct over the pre-existing struct , this way same struct can be used in gettestconfig as well as `test`

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

File: cmd/test.go (hunk 12 above)
Comment (Akash-Singh04) [reply cmd/test.go:258]: We are using TestConfigOptions only once here

File: pkg/models/testconfigoptions.go (hunk 13 above)
Comment (Akash-Singh04) [reply pkg/models/testconfigoptions.go:22]: The two structs TestOptions and TestConfigOptions have some variances in properties, which will be needed to be set to NULL if we combine make a struct over the pre existing one if i am not wrong

File: pkg/service/record/service.go (hunk 11 above)
Comment (Akash-Singh04) [reply pkg/service/record/service.go]: Added And Refactored

File: pkg/models/recordoptions.go
Comment (Akash-Singh04) [reply pkg/models/recordoptions.go:1]: Refactored

File: cmd/record.go (hunk 10 above)
Comment (Akash-Singh04) [reply cmd/record.go]: This is viable for RecordOptions, Implemented

File: cmd/test.go (hunk 12 above)
Comment (shivamsouravjha) [reply cmd/test.go:258]: I checked the `getTestConfig` function compared it to `StartTest` and only one of field is additional : `configPath`.

So a superset called testConfig can be made inside ~
`
testConfig {
testOption
configpath
}
`

**Review by shivamsouravjha — CHANGES_REQUESTED** [review]

**Comment by Akash-Singh04** [comment]
Closing this PR as the desired outcomes were not acheived.
